File: app/src/main/java/io/recebimento/network/SessionManager.kt
package io.recebimento.network

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.util.Date

class SessionManager(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences("RecebimentoPrefs", Context.MODE_PRIVATE)

    companion object {
        private const val TAG = "SessionManager"
        private const val KEY_BEARER_TOKEN = "BEARER_TOKEN"
        private const val KEY_TOKEN_EXPIRY = "TOKEN_EXPIRY"
        private const val KEY_USER_EMAIL = "USER_EMAIL"
        private const val KEY_USER_NAME = "USER_NAME"
        private const val KEY_USER_STORE = "USER_STORE"
        private const val KEY_REMEMBER_LOGIN = "REMEMBER_LOGIN"
        private const val KEY_MS_PASSWORD = "MS_PASSWORD"
        private const val KEY_BRLOG_REFRESH_TOKEN = "BRLOG_REFRESH_TOKEN"
        private const val KEY_BRLOG_PROGRESS = "BRLOG_PROGRESS"
        private const val KEY_BRLOG_NOTAS = "BRLOG_NOTAS"
        private const val KEY_BRLOG_COD_EMPRESA_USUARIO = "BRLOG_COD_EMPRESA_USUARIO"
        private const val TOKEN_EXPIRY_DAYS = 14
    }

    fun saveBrlogRefreshToken(refreshToken: String) {
        prefs.edit().putString(KEY_BRLOG_REFRESH_TOKEN, refreshToken).apply()
    }

    fun getBrlogRefreshToken(): String? = prefs.getString(KEY_BRLOG_REFRESH_TOKEN, null)

    fun saveBrlogProgress(progresso: Map<String, Double>) {
        try {
            prefs.edit().putString(KEY_BRLOG_PROGRESS, Gson().toJson(progresso)).apply()
        } catch (e: Exception) {
            Log.e(TAG, "saveBrlogProgress: Erro", e)
        }
    }

    fun getBrlogProgress(): Map<String, Double> {
        return try {
            val json = prefs.getString(KEY_BRLOG_PROGRESS, null) ?: return emptyMap()
            val type = object : TypeToken<Map<String, Double>>() {}.type
            Gson().fromJson<Map<String, Double>>(json, type) ?: emptyMap()
        } catch (e: Exception) {
            Log.e(TAG, "getBrlogProgress: Erro", e)
            emptyMap()
        }
    }

    fun saveBrlogNotas(notas: List<BrasilRiskNota>) {
        try {
            prefs.edit().putString(KEY_BRLOG_NOTAS, Gson().toJson(notas)).apply()
        } catch (e: Exception) {
            Log.e(TAG, "saveBrlogNotas: Erro", e)
        }
    }

    fun getBrlogNotas(): List<BrasilRiskNota> {
        return try {
            val json = prefs.getString(KEY_BRLOG_NOTAS, null) ?: return emptyList()
            val type = object : TypeToken<List<BrasilRiskNota>>() {}.type
            Gson().fromJson<List<BrasilRiskNota>>(json, type) ?: emptyList()
        } catch (e: Exception) {
            Log.e(TAG, "getBrlogNotas: Erro", e)
            emptyList()
        }
    }

    fun saveBrlogCodEmpresaUsuario(codEmpresaUsuario: Int) {
        prefs.edit().putInt(KEY_BRLOG_COD_EMPRESA_USUARIO, codEmpresaUsuario).apply()
    }

    fun getBrlogCodEmpresaUsuario(): Int = prefs.getInt(KEY_BRLOG_COD_EMPRESA_USUARIO, 0)

    fun saveToken(token: String) {
        try {
            Log.d(TAG, "saveToken: Salvando token (${token.take(30)}...)")
            
            val editor = prefs.edit()
            editor.putString(KEY_BEARER_TOKEN, token)
            
            val expiry = System.currentTimeMillis() + (TOKEN_EXPIRY_DAYS * 24 * 60 * 60 * 1000L)
            editor.putLong(KEY_TOKEN_EXPIRY, expiry)
            editor.putBoolean(KEY_REMEMBER_LOGIN, true)
            editor.apply()
            
            Log.d(TAG, "saveToken: Token salvo com sucesso!")
        } catch (e: Exception) {
            Log.e(TAG, "saveToken: Erro ao salvar token", e)
        }
    }

    fun getToken(): String? {
        return try {
            val token = prefs.getString(KEY_BEARER_TOKEN, null)
            
            if (token.isNullOrEmpty()) {
                Log.d(TAG, "getToken: Nenhum token encontrado")
                return null
            }
            
            if (isTokenExpired()) {
                Log.e(TAG, "getToken: Token EXPIRADO! Removendo...")
                clearToken()
                return null
            }
            
            Log.d(TAG, "getToken: Token VÁLIDO")
            token
            
        } catch (e: Exception) {
            Log.e(TAG, "getToken: Erro ao obter token", e)
            null
        }
    }

    fun isLoggedIn(): Boolean {
        return try {
            val remember = prefs.getBoolean(KEY_REMEMBER_LOGIN, false)
            val token = getToken()
            remember && !token.isNullOrEmpty()
        } catch (e: Exception) {
            Log.e(TAG, "isLoggedIn: Erro", e)
            false
        }
    }

    private fun isTokenExpired(): Boolean {
        val expiry = prefs.getLong(KEY_TOKEN_EXPIRY, 0)
        if (expiry == 0L) return false
        return System.currentTimeMillis() > expiry
    }

    fun clearToken() {
        try {
            prefs.edit().remove(KEY_BEARER_TOKEN).apply()
            prefs.edit().remove(KEY_TOKEN_EXPIRY).apply()
            prefs.edit().putBoolean(KEY_REMEMBER_LOGIN, false).apply()
        } catch (e: Exception) {
            Log.e(TAG, "clearToken: Erro ao remover token", e)
        }
    }

    fun saveUserInfo(email: String, name: String, store: String) {
        try {
            prefs.edit().apply {
                putString(KEY_USER_EMAIL, email)
                putString(KEY_USER_NAME, name)
                putString(KEY_USER_STORE, store)
            }.apply()
        } catch (e: Exception) {
            Log.e(TAG, "saveUserInfo: Erro ao salvar info", e)
        }
    }

    fun getUserEmail(): String? = prefs.getString(KEY_USER_EMAIL, null)
    
    fun getUserStore(): String? = prefs.getString(KEY_USER_STORE, "L291")

    fun saveCredentials(email: String, password: String) {
        try {
            prefs.edit().apply {
                putString(KEY_USER_EMAIL, email)
                putString(KEY_MS_PASSWORD, password)
            }.apply()
            Log.d(TAG, "saveCredentials: Credenciais salvas")
        } catch (e: Exception) {
            Log.e(TAG, "saveCredentials: Erro", e)
        }
    }

    fun getSavedPassword(): String? {
        return try {
            prefs.getString(KEY_MS_PASSWORD, null)
        } catch (e: Exception) {
            Log.e(TAG, "getSavedPassword: Erro", e)
            null
        }
    }

    fun hasSavedCredentials(): Boolean {
        val email = getUserEmail()
        val senha = getSavedPassword()
        return !email.isNullOrBlank() && !senha.isNullOrBlank()
    }

    fun clearCredentials() {
        try {
            prefs.edit().remove(KEY_MS_PASSWORD).apply()
            Log.d(TAG, "clearCredentials: Credenciais removidas")
        } catch (e: Exception) {
            Log.e(TAG, "clearCredentials: Erro", e)
        }
    }

    fun clearAll() {
        try {
            prefs.edit().clear().apply()
        } catch (e: Exception) {
            Log.e(TAG, "clearAll: Erro ao limpar preferências", e)
        }
    }
}

File: flutter/lib/network/session_manager.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../../models/brasil_risk.dart';

class SessionManager {
  static const String _prefsName = "RecebimentoPrefs";
  static const String _keyBearerToken = "BEARER_TOKEN";
  static const String _keyTokenExpiry = "TOKEN_EXPIRY";
  static const String _keyUserEmail = "USER_EMAIL";
  static const String _keyUserName = "USER_NAME";
  static const String _keyUserStore = "USER_STORE";
  static const String _keyRememberLogin = "REMEMBER_LOGIN";
  static const String _keyMsPassword = "MS_PASSWORD";
  static const String _keyBrlogRefreshToken = "BRLOG_REFRESH_TOKEN";
  static const String _keyBrlogProgress = "BRLOG_PROGRESS";
  static const String _keyBrlogNotas = "BRLOG_NOTAS";
  static const String _keyBrlogCodEmpresaUsuario = "BRLOG_COD_EMPRESA_USUARIO";
  static const int _tokenExpiryDays = 14;

  final SharedPreferences _prefs;

  SessionManager(this._prefs);

  static Future<SessionManager> create() async {
    final prefs = await SharedPreferences.getInstance();
    return SessionManager(prefs);
  }

  // ---------- BRLog ----------
  void saveBrlogRefreshToken(String token) =>
      _prefs.setString(_keyBrlogRefreshToken, token);

  String? getBrlogRefreshToken() => _prefs.getString(_keyBrlogRefreshToken);

  void saveBrlogProgress(Map<String, double> progresso) =>
      _prefs.setString(_keyBrlogProgress, jsonEncode(progresso));

  Map<String, double> getBrlogProgress() {
    final json = _prefs.getString(_keyBrlogProgress);
    if (json == null) return {};
    try {
      final map = jsonDecode(json) as Map<String, dynamic>;
      return map.map((k, v) => MapEntry(k, (v as num).toDouble()));
    } catch (_) {
      return {};
    }
  }

  void saveBrlogNotas(List<BrasilRiskNota> notas) =>
      _prefs.setString(_keyBrlogNotas, jsonEncode(notas.map((n) => _notaToMap(n)).toList()));

  List<BrasilRiskNota> getBrlogNotas() {
    final json = _prefs.getString(_keyBrlogNotas);
    if (json == null) return [];
    try {
      final list = jsonDecode(json) as List;
      return list.map((e) => BrasilRiskNota.fromJson(e)).toList();
    } catch (_) {
      return [];
    }
  }

  void saveBrlogCodEmpresaUsuario(int cod) =>
      _prefs.setInt(_keyBrlogCodEmpresaUsuario, cod);

  int getBrlogCodEmpresaUsuario() =>
      _prefs.getInt(_keyBrlogCodEmpresaUsuario) ?? 0;

  // ---------- Token principal ----------
  void saveToken(String token) {
    _prefs.setString(_keyBearerToken, token);
    final expiry = DateTime.now()
        .add(const Duration(days: _tokenExpiryDays))
        .millisecondsSinceEpoch;
    _prefs.setInt(_keyTokenExpiry, expiry);
    _prefs.setBool(_keyRememberLogin, true);
  }

  void saveTokenWithExpiry(String token, {int? expiryEpochSeconds}) {
    final expiry = (expiryEpochSeconds != null
            ? expiryEpochSeconds * 1000
            : DateTime.now().millisecondsSinceEpoch +
                (_tokenExpiryDays * 24 * 60 * 60 * 1000))
        .clamp(
            DateTime.now().millisecondsSinceEpoch,
            DateTime.now()
                .add(const Duration(days: 365))
                .millisecondsSinceEpoch);
    _prefs.setString(_keyBearerToken, token);
    _prefs.setInt(_keyTokenExpiry, expiry);
    _prefs.setBool(_keyRememberLogin, true);
  }

  String? getToken() {
    final token = _prefs.getString(_keyBearerToken);
    if (token == null || token.isEmpty) return null;
    if (isTokenExpired()) {
      clearToken();
      return null;
    }
    return token;
  }

  bool isLoggedIn() {
    final remember = _prefs.getBool(_keyRememberLogin) ?? false;
    final token = getToken();
    return remember && token != null && token.isNotEmpty;
  }

  bool isTokenExpired() {
    final expiry = _prefs.getInt(_keyTokenExpiry) ?? 0;
    if (expiry == 0) return false;
    return DateTime.now().millisecondsSinceEpoch > expiry;
  }

  void clearToken() {
    _prefs.remove(_keyBearerToken);
    _prefs.remove(_keyTokenExpiry);
    _prefs.setBool(_keyRememberLogin, false);
  }

  void saveUserInfo(String email, String name, String store) {
    _prefs.setString(_keyUserEmail, email);
    _prefs.setString(_keyUserName, name);
    _prefs.setString(_keyUserStore, store);
  }

  String? getUserEmail() => _prefs.getString(_keyUserEmail);
  String getUserStore() => _prefs.getString(_keyUserStore) ?? "L291";

  void saveCredentials(String email, String password) {
    _prefs.setString(_keyUserEmail, email);
    _prefs.setString(_keyMsPassword, password);
  }

  String? getSavedPassword() => _prefs.getString(_keyMsPassword);

  bool hasSavedCredentials() {
    final email = getUserEmail();
    final senha = getSavedPassword();
    return email != null &&
        email.isNotEmpty &&
        senha != null &&
        senha.isNotEmpty;
  }

  void clearCredentials() => _prefs.remove(_keyMsPassword);

  void clearAll() => _prefs.clear();
}

Map<String, dynamic> _notaToMap(BrasilRiskNota n) => {
      'CodPedido': n.codPedido,
      'CodPedidoDestino': n.codPedidoDestino,
      'CodPedidoOcorrenciaTipo': n.codPedidoOcorrenciaTipo,
      'CodEmpresaUsuario': n.codEmpresaUsuario,
      'NumeroViagem': n.numeroViagem,
      'ProgressoViagem': n.progressoViagem,
      'CodStatusDaEntrega': n.codStatusDaEntrega,
      'StatusDaViagem': n.statusDaViagem,
      'DataDeSaida': n.dataDeSaida,
      'DataIniciado': n.dataIniciado,
      'DistanciaRestante': n.distanciaRestante,
      'DistanciaTotalPrevista': n.distanciaTotalPrevista,
      'DistanciaPercorrida': n.distanciaPercorrida,
      'NomeMotorista': n.nomeMotorista,
      'Placa': n.placa,
      'Carreta': n.carreta,
      'PrevisaoChegada': n.previsaoChegada,
      'PrevisaoChegadaRecalculada': n.previsaoChegadaRecalculada,
      'DtPrevisaoEntrega': n.dtPrevisaoEntrega,
      'DataChegadaOrigemRecalculada': n.dataChegadaOrigemRecalculada,
      'DataConclusao': n.dataConclusao,
      'DataEntrega': n.dataEntrega,
      'DataDeEntregaNF': n.dataDeEntregaNF,
      'Status': n.status,
    };
